### PR TITLE
added single message pickup feature

### PR DESCRIPTION
**Pull request by dgunseli**
Added single message pickup feature for querying messages with a message_id value. This feature might be useful for push notification communication because of the agent message sizes can be exceed data limits for push notifications very easily.

### CODE REVIEW

**Comment by dhh1128** [comment]
I like the latest update.

@TelegramSam , as a person who's advocated pickup often, what feedback do you have?

**Comment by dgunseli** [comment]
Should there be any "register" function? I guess all the messages of the all users should not be stored and provided by this function. User should request for registering themselves for using message pickup feature.

File: features/0212-pickup/README.md [hunk 1]
```diff
@@ -70,7 +70,57 @@ A message that contains multiple waiting messages.
 {
     "@id": "123456781",
     "@type": "did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/messagepickup/1.0/batch",
-    "messages": []
```
Comment (TelegramSam) [inline features/0212-pickup/README.md]: How do you feel about `messages~attach` as the attribute/attachment method?

File: features/0212-pickup/README.md [hunk 2]
```diff
@@ -70,7 +70,57 @@ A message that contains multiple waiting messages.
...
+}
+```
+### Message Query With Message Id List
+A request to read single or multiple messages with a message message id array.
+```json=
+{
+    "@id": "123456781",
+    "@type": "did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/messagepickup/1.0/single_pickup",
+    "message_id": "06ca25f6-d3c5-48ac-8eee-1a9e29120c31"
+    "@type": "did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/messagepickup/1.0/list_pickup",
```
Comment (TelegramSam) [inline features/0212-pickup/README.md]: This example is malformed. It seems like you intended to include two examples that got combined by mistake.

**Comment by TelegramSam** [comment]
> @TelegramSam and @dhh1128 , should there be any "register" function? I guess all the messages of the all users should not be stored and provided by this function. User should request for registering themselves for using message pickup feature.

Users can only pick up messages intended for them. This is verified either by the key used for the pickup request, or via keys registered for routing as part of that protocol.

File: features/0212-pickup/README.md (hunk 1 above)
Comment (dhh1128) [reply features/0212-pickup/README.md]: +1 to @TelegramSam 's suggestion.

**Comment by dgunseli** [comment]
> > @TelegramSam and @dhh1128 , should there be any "register" function? I guess all the messages of the all users should not be stored and provided by this function. User should request for registering themselves for using message pickup feature.
> 
> Users can only pick up messages intended for them. This is verified either by the key used for the pickup request, or via keys registered for routing as part of that protocol.

I mean, user will be able to query their stored messages with id or gather with batch operation. But, as a mediator, system shouldn't store messages of all the users. I think each user should register themselves for pickup feature and then the system should start to store their messages and provide with pickup feature.

**Comment by TelegramSam** [comment]
The registration option seems well covered by [RFC 0211](https://github.com/hyperledger/aries-rfcs/blob/master/features/0211-route-coordination/README.md). Was there something different than you imagined?

**Comment by dgunseli** [comment]
@TelegramSam and @dhh1128 ,
Should there be any handshaking protocol? Let's assume this scenario:

Edge agent queries messages with batch query and mediator returns messages but an error occurs while edge agent is processing some of the messages. When edge agent queries its messages again with batch query, messages whose process not completed by the edge agent should be return by the mediator. So, edge agent should send "ack" after it completes process each messages.

**Comment by dhh1128** [comment]
>Should there be any handshaking protocol? Let's assume this scenario...

You are now describing something that is very, very close to the IMAP protocol. I think we could learn some things from it. It explicitly separates pickup from the release-from-server-storage process. That is, you fetch messages. Entirely separate from that, you can delete messages held on an IMAP server. You might delete messages you've fetched, or messages you've not fetched. They are just two separate operations.

Although I think the above may be useful input, I confess that as I see us inventing something so similar to IMAP, I am becoming uncomfortable about the feature. Maybe we should abandon it entirely, and simply tell agents who want this behavior to use IMAP as their transport. All this complexity would vanish, and there would be a lot less code to write.

**Comment by dgunseli** [comment]
> > Should there be any handshaking protocol? Let's assume this scenario...
> 
> You are now describing something that is very, very close to the IMAP protocol. I think we could learn some things from it. It explicitly separates pickup from the release-from-server-storage process. That is, you fetch messages. Entirely separate from that, you can delete messages held on an IMAP server. You might delete messages you've fetched, or messages you've not fetched. They are just two separate operations.
> 
> Although I think the above may be useful input, I confess that as I see us inventing something so similar to IMAP, I am becoming uncomfortable about the feature. Maybe we should abandon it entirely, and simply tell agents who want this behavior to use IMAP as their transport. All this complexity would vanish, and there would be a lot less code to write.

You right @dhh1128, it is close to IMAP, but defining a standard for message storing and forwarding  mechanism would be very useful for mediator approach and also this approach seems like easy for integrating edge agents, isn't it?

**Comment by dhh1128** [comment]
@dgunseli : : What would you like to do about this PR? I am trying not to let it languish without attention...

**Comment by denizgunseli** [comment]
@dhh1128 and @TelegramSam,

All requests were done. Can you review the changes?

**Comment by dhh1128** [comment]
I am okay with merging this. @TelegramSam ?

**Comment by TelegramSam** [comment]
@dhh1128 can you approve the requested changes?

**Pull request merged**